### PR TITLE
Exposing native memory usage in MemberState

### PR DESCRIPTION
**Pull request by erosb**
Exposes native memory usage data to MC about the member, and also about caches, maps, and near-caches. Respective `nativeMemoryUsed` booleans added to `*Stats` classes.

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java [hunk 1]
```diff
@@ -650,15 +653,26 @@ public void deregisterAllListener(String cacheNameWithPrefix) {
...
+            }
+            stats.put(cacheName, localCacheStats);
         }
         return stats;
     }
 
     @Override
     public CacheOperationProvider getCacheOperationProvider(String cacheNameWithPrefix, InMemoryFormat inMemoryFormat) {
-        if (InMemoryFormat.NATIVE.equals(inMemoryFormat)) {
+        if (NATIVE.equals(inMemoryFormat) && !BuildInfoProvider.getBuildInfo().isEnterprise()) {
```
Comment (emre-aydin) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java]: What does this change have to do with sending stats to MC?

File: hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java [hunk 2]
```diff
@@ -650,15 +653,26 @@ public void deregisterAllListener(String cacheNameWithPrefix) {
...
         for (Map.Entry<String, CacheStatisticsImpl> entry : statistics.entrySet()) {
-            stats.put(entry.getKey(), new LocalCacheStatsImpl(entry.getValue()));
+            String cacheName = entry.getKey();
+            LocalCacheStatsImpl localCacheStats = new LocalCacheStatsImpl(entry.getValue());
+            CompletableFuture<CacheConfig> configFuture = configs.get(cacheName);
+            if (configFuture.isDone()) {
+                try {
+                    localCacheStats.setNativeMemoryUsed(configFuture.get().getInMemoryFormat() == NATIVE);
+                } catch (InterruptedException | ExecutionException e) {
+                    e.printStackTrace();
```
Comment (emre-aydin) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java]: Should be logged, not printed to standard output.

File: hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java [hunk 3]
```diff
@@ -650,15 +653,26 @@ public void deregisterAllListener(String cacheNameWithPrefix) {
...
         Map<String, LocalCacheStats> stats = createHashMap(statistics.size());
+
         for (Map.Entry<String, CacheStatisticsImpl> entry : statistics.entrySet()) {
-            stats.put(entry.getKey(), new LocalCacheStatsImpl(entry.getValue()));
+            String cacheName = entry.getKey();
+            LocalCacheStatsImpl localCacheStats = new LocalCacheStatsImpl(entry.getValue());
+            CompletableFuture<CacheConfig> configFuture = configs.get(cacheName);
+            if (configFuture.isDone()) {
+                try {
+                    localCacheStats.setNativeMemoryUsed(configFuture.get().getInMemoryFormat() == NATIVE);
```
Comment (emre-aydin) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java]: Same as above, we either should send the in-memory format or rename to `persistentMemoryUsed` and set it to `true` when persistent memory is enabled.

File: hazelcast/src/main/java/com/hazelcast/internal/monitor/LocalCacheStats.java [hunk 4]
```diff
@@ -121,4 +121,6 @@
      */
     float getAverageRemoveTime();
 
+    boolean isNativeMemoryUsed();
```
Comment (emre-aydin) [inline hazelcast/src/main/java/com/hazelcast/internal/monitor/LocalCacheStats.java]: A javadoc would be welcome.

File: hazelcast/src/main/java/com/hazelcast/map/LocalMapStats.java [hunk 5]
```diff
@@ -274,4 +274,6 @@
      */
     Map<String, LocalIndexStats> getIndexStats();
 
+    boolean isNativeMemoryUsed();
```
Comment (emre-aydin) [inline hazelcast/src/main/java/com/hazelcast/map/LocalMapStats.java]: We need javadoc here. When we add this method here, it also becomes public API, not only used by MC.

File: hazelcast/src/main/java/com/hazelcast/nearcache/NearCacheStats.java [hunk 6]
```diff
@@ -129,4 +129,6 @@
      * @return the failure reason of the last Near Cache persistence
      */
     String getLastPersistenceFailure();
+
+    boolean isNativeMemoryUsed();
```
Comment (emre-aydin) [inline hazelcast/src/main/java/com/hazelcast/nearcache/NearCacheStats.java]: Javadoc.

File: hazelcast/src/test/java/com/hazelcast/internal/management/TimedMemberStateFactoryTest.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    public void before() {
+        System.setProperty(HAZELCAST_INTERNAL_OVERRIDE_ENTERPRISE, "true");
+    }
+
+    @After
+    public void after() {
+        if (instance != null) {
+            instance.shutdown();
+        }
+        System.setProperty(HAZELCAST_INTERNAL_OVERRIDE_ENTERPRISE, "");
```
Comment (emre-aydin) [inline hazelcast/src/test/java/com/hazelcast/internal/management/TimedMemberStateFactoryTest.java]: Nit: you could use `System.clearProperty(...)`.

File: hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java (hunk 1 above)
Comment (erosb) [reply hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java]: Without this line the `TimedMemberStateFactoryTest#nativeMemoryEnabledForCache()` test fails with an exception with this stacktrace:
```
java.lang.IllegalArgumentException: Native memory is available only in Hazelcast Enterprise.Make sure you have Hazelcast Enterprise JARs on your classpath!

	at com.hazelcast.cache.impl.AbstractCacheService.getCacheOperationProvider(AbstractCacheService.java:677)
	at com.hazelcast.cache.impl.CacheProxySupport.<init>(CacheProxySupport.java:121)
	at com.hazelcast.cache.impl.CacheProxy.<init>(CacheProxy.java:95)
	at com.hazelcast.cache.impl.AbstractCacheService.createDistributedObject(AbstractCacheService.java:270)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyRegistry.doCreateProxy(ProxyRegistry.java:219)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyRegistry.createProxy(ProxyRegistry.java:211)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyRegistry.getOrCreateProxyFuture(ProxyRegistry.java:178)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyRegistry.getOrCreateProxy(ProxyRegistry.java:159)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyServiceImpl.getDistributedObject(ProxyServiceImpl.java:148)
	at com.hazelcast.instance.impl.HazelcastInstanceImpl.getDistributedObject(HazelcastInstanceImpl.java:352)
	at com.hazelcast.cache.impl.HazelcastInstanceCacheManager.getCacheByFullName(HazelcastInstanceCacheManager.java:51)
	at com.hazelcast.cache.impl.HazelcastInstanceCacheManager.getCache(HazelcastInstanceCacheManager.java:45)
	at com.hazelcast.internal.management.TimedMemberStateFactoryTest.nativeMemoryEnabledForCache(TimedMemberStateFactoryTest.java:94)
```

So I was looking for a workaround to overcome this hard check. Previously I saw that the `ConfigValidator#checkNotNativeWhenOpenSource()` method (which runs for every map and cache config) uses this `BuildInfo#isEnterprise()` flag for validation, and this flag is controllable using the `HAZELCAST_INTERNAL_OVERRIDE_ENTERPRISE` system property (that's why we toggle it in the `@Before` and the `@After` methods). So I needed to work this hard check around and I thought it makes sense to stick to the approach that is already utilized in an other part of the codebase (config validation).

Comment (emre-aydin) [reply hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java]: Thanks for the detailed explanation !

File: hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java (hunk 3 above)
Comment (erosb) [reply hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java]: The problem is that at this point we only know that native memory is used, which is not necessarily persistent memory. So I took the approach of sending the in-memory format. So on MC side we will have to use both the root-level `TimedMemberState#persistendMemoryUsed` _and_ the `#inMemoryFormat` fields of data structures when we determine if the data structure uses optane memory or not.

**Comment by erosb** [comment]
Hello @emre-aydin thank you for the comments, now this is ready for re-review.

File: hazelcast/src/main/java/com/hazelcast/config/MapConfig.java [hunk 8]
```diff
@@ -955,4 +957,13 @@ public void readData(ObjectDataInput in) throws IOException {
         eventJournalConfig = in.readObject();
         metadataPolicy = MetadataPolicy.getById(in.readShort());
     }
+
+    public LocalMapStatsImpl initMapStats() {
```
[the remark this answers is not in the document]
Comment (erosb) [reply hazelcast/src/main/java/com/hazelcast/config/MapConfig.java]: I moved the initialization into a static method of the LocalMapStats class.

**Comment by Holmistr** [comment]
run-lab-run

**Comment by Holmistr** [comment]
run-lab-run

**Comment by Holmistr** [comment]
run-lab-run

**Comment by Holmistr** [comment]
run-lab-run

**Comment by Holmistr** [comment]
run-lab-run

**Comment by erosb** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/map/impl/LocalMapStatsProvider.java [hunk 9]
```diff
@@ -152,7 +155,7 @@ private void addStatsOfNoDataIncludedMaps(Map statsPerMap) {
         Collection<String> mapNames = proxyService.getDistributedObjectNames(SERVICE_NAME);
         for (String mapName : mapNames) {
             if (!statsPerMap.containsKey(mapName)) {
-                statsPerMap.put(mapName, EMPTY_LOCAL_MAP_STATS);
+                statsPerMap.put(mapName, initializeLocalMapStats(mapName));
```
Comment (puzpuzpuz) [inline hazelcast/src/main/java/com/hazelcast/map/impl/LocalMapStatsProvider.java]: I wonder if this change could affect performance somehow? Before a single object was reused for empty maps and now it's allocated per each map.

File: hazelcast/src/main/java/com/hazelcast/internal/monitor/LocalCacheStats.java [hunk 10]
```diff
@@ -121,4 +122,11 @@
      */
     float getAverageRemoveTime();
 
+    /**
+     * Returns the in-memory format used to represent the contents of the cache on the member.
+     *
+     * @return the in-memory format used to represent the contents of the cache on the member
+     */
+    InMemoryFormat getInMemoryFormat();
```
[the remark this answers is not in the document]
Comment (emre-aydin) [reply hazelcast/src/main/java/com/hazelcast/internal/monitor/LocalCacheStats.java]: I think we should leave this info out of listing pages. We won't be able to support it when we switch to metrics.

[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply hazelcast/src/main/java/com/hazelcast/internal/monitor/LocalCacheStats.java]: Exactly. I also thought about availability of this data in metrics.

**Comment by puzpuzpuz** [comment]
@erosb 
Probably this PR should be closed, right?

**Pull request closed without merging**